docs/proposals/20231122-automate-aks-features.md

```diff
@@ -0,0 +1,397 @@
+version of CAPZ suddenly starts constructing invalid ASO resources and user intervention is required to
+perform the conversion manually.
+
+While it couples CAPZ to one ASO API version, this approach allows CAPZ to move a more calculated pace with
+regards to AKS API versions the way it's done today. This also narrows CAPZ's scope of responsibility which
+reduces CAPZ's exposure to potential incompatibilities with certain ASO API versions.
+
+Regarding ClusterClass, this option functions the same as [Option 2] or [Option 3], where all ASO fields can
+be defined in a template. This option opens up an additional safeguard though, where webhooks could flag
+fields which should not be defined in a template. Similar webhook checks would be less practical in the other
+options where CAPZ would need to be aware of the set of disallowed fields for each ASO API version that a user
+could use.
+
+Similarly, CAPZ's webhooks are also better able to validate and default the ASO configuration and ensure
+fields like ManagedCluster's `spec.owner` that should not be modified by users are set correctly.
+
+#### Option 5: No change: CAPZ resource evolution proceeds the way it currently does
+
+This method describes not making any changes to how the CAPZ API is generally structured for AKS resources.
+CAPZ API types will continue to be curated manually without inheriting anything from the ASO API.
+
+Benefits of continuing on our current path include:
+- Familiarity with the existing pattern by users and contributors
+- Zero up-front cost to implement or transition to a new pattern
+- No requirement for users to have ASO knowledge
+- Greater freedom to change API implementations which we've recently leveraged to transition between the older
+  and newer Azure SDKs and to ASO.
+
+#### Option 6: Generate CAPZ code equivalent to what's added manually today
+
+For this option, a new code generation pipeline would be created to automatically scaffold the code that is
+currently manually written to expose additional AKS API fields to CAPZ.
+
+Once implemented, this method would drastically reduce the amount of developer effort to get started adding
+new AKS features. There would continue to be some amount of ongoing cost though to identify and handle nuances
+for each feature though, which may include issues outside of CAPZ's control like AKS API quirks.
+
+The main drawback of this approach from a developer perspective is the up-front effort to implement and
+ongoing cost to maintain the code generation itself. The existing AKS feature set exposed by CAPZ provides a
+decent foundation that can help identify regressions by testing the pipeline against features already
+implemented and tested. ASO also already implements a full code generation pipeline to transform Azure API
+specs into Kubernetes resource definitions, so some of that could possibly be reused by CAPZ.
```

Comment:
One thing I like about this option is that we could start using it immediately to bootstrap a new (e.g., v2beta1) API for AzureManagedCluster, AzureManagedControlPlane, and AzureManagedMachinePool. Because that API doesn't exist at all we could poke the automation until we reliably get it to build that new API for us automatically.

Reply:
The way I imagine it is that every versioned set of ASO resources that CAPZ consumes renders the equivalent CAPZ code surface are w/ 100% feature completeness, and so we have a nice long-term story where the next ASO version we consume yields the new stuff automatically.